cnmfpy/algs/chals.py

```diff
@@ -0,0 +1,101 @@
+import numpy as np  # Linear algebra
+import numpy.linalg as la
+
+from tqdm import trange
+
+from ..conv import tensor_conv, shift_and_stack, hunfold
+from ..optimize import compute_loss, renormalize
+
+# TODO make EPSILON universal
+EPSILON = np.finfo(np.float).eps
+
+
+def fit_chals(data, model):
+    N, T = data.shape
+
+    # initial loss
+    model.loss_hist = [compute_loss(data, model.W, model.H)]
+
+    itr = 0
+    for itr in trange(model.n_iter_max):
+
+        if (np.isnan(model.W).any()):
+            raise Exception('W has Nans!!')
+
+        if (np.isnan(model.H).any()):
+            raise Exception('H has NANs!!')
+
+        # shift factors
+        # if ((itr % 5 == 0) and (model.n_iter_max - itr > 15)):
+            # model.W, model.H = shift_factors(model.W, model.H)
+
+        for k in range(model.n_components):
+            update_W_component(model.W, model.H, data, k)
+            update_H_component(model.W, model.H, data, k)
+
+        # Update loss history and renormalize H
+        model.loss_hist += [compute_loss(data, model.W, model.H)]
+        model.W, model.H = renormalize(model.W, model.H)
+
+        # check convergence
+        prev_loss, new_loss = model.loss_hist[-2:]
+        if (np.abs(prev_loss - new_loss) < model.tol):
+            break
+
+
+def update_W_col(W, H, X, k, l):
```

Comment:
I'd vote to get rid of this function and just make it part of `update_W_component`